ksmt-symfpu/src/main/kotlin/org/ksmt/symfpu/SymFPUModel.kt

```diff
@@ -0,0 +1,179 @@
+package org.ksmt.symfpu
+
+import org.ksmt.KContext
+import org.ksmt.decl.KDecl
+import org.ksmt.expr.KArrayConst
+import org.ksmt.expr.KArrayLambdaBase
+import org.ksmt.expr.KArrayStoreBase
+import org.ksmt.expr.KConst
+import org.ksmt.expr.KExpr
+import org.ksmt.expr.KUninterpretedSortValue
+import org.ksmt.solver.KModel
+import org.ksmt.solver.model.KModelEvaluator
+import org.ksmt.solver.model.KModelImpl
+import org.ksmt.sort.KArraySortBase
+import org.ksmt.sort.KFpSort
+import org.ksmt.sort.KSort
+import org.ksmt.sort.KUninterpretedSort
+import org.ksmt.symfpu.ArraysTransform.Companion.mkAnyArrayLambda
+import org.ksmt.symfpu.ArraysTransform.Companion.mkAnyArraySelect
+import org.ksmt.symfpu.ArraysTransform.Companion.mkAnyArrayStore
+import org.ksmt.utils.cast
+import org.ksmt.utils.uncheckedCast
+
+class SymFPUModel(private val kModel: KModel, val ctx: KContext, val transformer: FpToBvTransformer) : KModel {
+    override val declarations: Set<KDecl<*>>
+        get() = kModel.declarations +
+            transformer.arraysTransform.mapFpToBvDeclImpl.keys -
+            transformer.arraysTransform.mapFpToBvDeclImpl.values.map { it.decl }.toSet()
+
+
+    override val uninterpretedSorts: Set<KUninterpretedSort>
+        get() = kModel.uninterpretedSorts
+
+
+    private val evaluatorWithModelCompletion by lazy { KModelEvaluator(ctx, this, isComplete = true) }
+    private val evaluatorWithoutModelCompletion by lazy { KModelEvaluator(ctx, this, isComplete = false) }
+    private val interpretations: MutableMap<KDecl<*>, KModel.KFuncInterp<*>> = hashMapOf()
+
+    override fun <T : KSort> eval(expr: KExpr<T>, isComplete: Boolean): KExpr<T> {
+        ctx.ensureContextMatch(expr)
+
+        val evaluator = if (isComplete) evaluatorWithModelCompletion else evaluatorWithoutModelCompletion
+        return evaluator.apply(expr)
+    }
+
+    private fun <T : KSort> getConst(decl: KDecl<T>): KExpr<*>? =
+        transformer.arraysTransform.mapFpToBvDeclImpl[decl]
+
+    override fun <T : KSort> interpretation(decl: KDecl<T>): KModel.KFuncInterp<T>? = with(ctx) {
+        ensureContextMatch(decl)
+        return interpretations.getOrPut(decl) {
+            if (!declContainsFp<T>(decl)) {
+                val interpretation = kModel.interpretation(decl)
+                if (interpretation != null) {
+                    return@getOrPut interpretation
+                } else return@with null
+            }
+
+            val const = getConst(decl) ?: return@with null
+            val eval = kModel.eval(const)
```

Comment:
It is better to use `kModel.interpretation` here. The reason is that interpretations can be partial instead of expressions. Also, interpretations allow us to avoid creation of complex expressions. For example, model of an uninterpreted function is usually looks like:
```
{x0 ->y0, x1 -> y1}
```
which is evaluated to 
```
(ite (= arg x0) y0 (ite (= arg x1) y1 <some random value>))
```